internal/mapdecode/hooks.go

```diff
@@ -175,3 +185,124 @@ func strconvHook(from, to reflect.Type, data reflect.Value) (reflect.Value, erro
 
 	return data, nil
 }
+
+// fieldHook applies the user-specified FieldHookFunc to all struct fields.
+func fieldHook(opts *options) reflectHook {
+	hook := opts.FieldHook
+	return func(from, to reflect.Type, data reflect.Value) (reflect.Value, error) {
+		if to.Kind() != reflect.Struct || from.Kind() != reflect.Map {
+			return data, nil
+		}
+
+		// We can only decode map[string]* and map[interface{}]* into structs.
+		if k := from.Key().Kind(); k != reflect.String && k != reflect.Interface {
+			return data, nil
+		}
+
+		// This map tracks type-changing updates to items in the map.
+		//
+		// If the source map has a rigid type for values (map[string]string
+		// rather than map[string]interface{}), we can't make replacements to
+		// values in-place if a hook changed the type of a value. So we will
+		// make a copy of the source map with a more liberal type and inject
+		// these updates into the copy.
+		updates := make(map[interface{}]interface{})
+
+		var errors []error
+		for i := 0; i < to.NumField(); i++ {
+			structField := to.Field(i)
+			if structField.PkgPath != "" {
+				// This field is not exported so we won't be able to decode
+				// into it.
+				continue
+			}
+
+			fieldName := structField.Name
+
+			// Field name override was specified.
+			tagParts := strings.Split(structField.Tag.Get(opts.TagName), ",")
+			if len(tagParts) > 0 && tagParts[0] != "" {
+				fieldName = tagParts[0]
+			}
+
+			// Get the value for this field from the source map, if any.
+			key := reflect.ValueOf(fieldName)
+			value := data.MapIndex(key)
+			if !value.IsValid() {
+				// Case-insensitive linear search if the name doesn't match
```

Comment:
Do we need a case insensitive search, I don't se anything in the mapstrucutre docs that says it does this